File: src/0031-Next-Permutation/0031.py
class Solution:
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        i = j = r = len(nums) - 1
        while i > 0 and nums[i] <= nums[i-1]:
            i -= 1
            
        i -= 1
        if i >= 0:
            while j >= i and nums[j] <= nums[i]:
                j -= 1

            nums[i], nums[j] = nums[j], nums[i]
            
        i += 1      
        while i < r:
            nums[i], nums[r] = nums[r], nums[i]
            i += 1
            r -= 1



if __name__ == "__main__":
    nums = [3,2,1]
    Solution().nextPermutation(nums)
    print(nums)

File: src/0031-Next-Permutation/0031.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

static int x = [](){std::ios::sync_with_stdio(false);cin.tie(0);return 0;}();
class Solution 
{
public:
    void nextPermutation(vector<int>& nums) 
    {
        int i = nums.size()-1, j = i;
        while (i > 0 and nums[i] <= nums[i-1]) --i;
        --i;
        if (i >= 0)
        {
            while (j >= i and nums[j] <= nums[i]) --j;
            int tmp = nums[i];
            nums[i] = nums[j]; nums[j] = tmp;
        }
        ++i;
        reverse(nums.begin()+i, nums.end());
    }
};
int main()
{
    vector<int> arr = { 3, 2, 2, 3};
    Solution().nextPermutation(arr);
    return 0;
}